SelectDropdown: cleanup code and migrate CSS to webpack

jsnajdr opened the pull request:
The `SelectDropdown` component accumulated a lot of cruft over the years, so I did a comprehensive cleanup as part of the CSS migration.

Please review commit by commit, I tried to keep them very small and explain everything.

Similar to `SegmentedControl` in #35051, I put the subcomponents as static properties on the main component:
```jsx
<SelectDropdown>
  <SelectDropdown.Item />
  <SelectDropdown.Label />
  <SelectDropdown.Separator />
</SelectDropdown>
```

jsnajdr committed:
SelectDropdown: replace methods bound in constructor with arrow fns

jsnajdr committed:
SelectDropdown: instanceId doesn't need to be in state and be used for keys

Changes `instanceId` to be an instance property assigned during construction. No need
to put it in state, as it never changes.

Also, don't use `instanceId` to create `key` values. These don't need to be globally unique.
Use them only for DOM element IDs that are used to link elements together with ARIA attributes.

jsnajdr committed:
SelectDropdown: remove props argument of getInitialSelectedItem

We can always read from `this.props`. Even in constructor, because we
called `super( props )`.

jsnajdr committed:
SelectDropdown: comment and improve the getInitialSelectedItem logic

Add comments about what the method does, and fix a bug where a separator
item (specified as `null` in `options`) can potentially crash the `find`.

jsnajdr committed:
SelectDropdown: toggle visibility of dropdown with visibility: hidden

Removes need for smart `tabIndex` logic: hidden elements are automatically
excluded from tab order.

`visibility: hidden`, as opposed to `display: none`, still takes the element
into account when doing layout. Therefore, the dropdown header will continue
to have the width of the widest option in the list.

jsnajdr committed:
SelectDropdown: use arrow functions where appropriate

jsnajdr committed:
SelectDropdown: use modern refs for focusing item links

Replace string ref with a modern one, and create a `focusLink` instance method
to provide API for the parent component that wants to set focus. Replace access
to internal `itemRef.refs.itemLink` property.

jsnajdr committed:
SelectDropdown: replace string ref to dropdownContainer with modern one

jsnajdr committed:
SelectDropdown: replace string itemRefs with modern ones

Use an array of refs to focusable items, set them with a callback, move the `refIndex`
increment inline to calling expression.

jsnajdr committed:
SelectDropdown: cloned children don't need a key

`key` is needed only when passing array of children. In this case, the children
are written as JSX and `key` doesn't need to be added on cloning and mapping 1:1.

jsnajdr committed:
SelectDropdown: only DropdownItem children need to be cloned and amended

Only `DropdownItem` needs a ref (it's focusable and the ref is used to move focus
on up and down keyboard navigation) and an `onClick` handler. Separator and Label
are neither focusable nor clickable.

jsnajdr committed:
SelectDropdown: ignore clicks on Label and Separator, add a11y role

Clicks on label and separator should not bubble to the parent element and
cause the dropdown to close. This patch adds a missing handler to the separator
component (label is already OK). Also adds an a11y role to specify that the element
is not interactive despite having an `onClick` handler.

jsnajdr committed:
SelectDropdown: simplify setting initial state

Can be done by assinging an instance property instead of full constructor.
Also, `getInitialSelectedItem` already handles the case where `options` prop
is not present or empty.

jsnajdr committed:
SelectDropdown: remove unneeded componentWillReceiveProps

Closing the popup when receiving new props doesn't seem to make much sense.
And the initial selected value is set only on initial mount and further changes
of the prop are ignored. That's the common behavior of initial-ish props on
uncontrolled components. For example, native `<input defaultValue="x" />` renders
input box with "x" and doesn't change the value on further rerenders with a different
prop.

jsnajdr committed:
SelectDropdown: no need to look at initial selected item in getSelectedText or Icon

In the `getSelectedText` and `getSelectedIcon` getters, the currently selected value
is always in `this.state.selected`. No need to default to the initial value.

Also, use `_.get` instead of `_.result`, as `icon` and `label` are not functions.

jsnajdr committed:
SelectDropdown: remove ESLint-offending and non-helpful JSDoc comment

jsnajdr committed:
SelectDropdown: fix a11y issues by adding roles and correct ARIA attrs

jsnajdr committed:
SelectDropdown: use the classNames helper better

jsnajdr committed:
SelectDropdown: use Lodash noop for event handler defaults

jsnajdr committed:
SelectDropdown: reorg the code that moves focus, use fewer lodash funcs

jsnajdr committed:
SelectDropdown: export Item, Separator and Label as main component fields

jsnajdr committed:
SelectDropdown: rework docs and devdocs example to use the SelectDropdown._ convention

jsnajdr committed:
SelectDropdown: migrate CSS to webpack

jsnajdr committed:
Update usages of SelectDropdown to use the SelectDropdown.Item convention

jsnajdr committed:
Remove import of SelectDropdown stylesheet from EditorPublishData

sgomes commented:
FYI @jsnajdr: I have an open PR atm which is touching the same component: #34989

sgomes commented:
@jsnajdr It looks like your PR supersedes mine, so I can probably close it. What do you think? Looks like we were working on the same thing at the same time 😄

jsnajdr commented:
> It looks like your PR supersedes mine, so I can probably close it.

I wanted to just migrate the CSS, but then got absorbed by making a lot if improvements in 22 commits 😄 

Looking at your PR, I'm doing a few things differently:
- I don't clear the `this.itemRefs` array on each render. Maybe I should do it?
- I don't check if a `ref.current` is set before calling `focus()` on it. I think they are always guaranteed to be set. They are accessed inside DOM handlers, which means the respective elements are always present.
- to focus the link inside the item, I created a `focusLink` instance method. There's a chain of two refs in action.

sgomes commented:
> Looking at your PR, I'm doing a few things differently:

I think your changes are generally better.

> * I don't clear the `this.itemRefs` array on each render. Maybe I should do it?

The reason I did this was to avoid keeping around references for longer than they should exist. If option `n` disappears and the total number of elements goes down, then it may remain in the array while the other indices get reassigned. It's probably not an issue at all in practice, but I figured I'd play it safe.

> * I don't check if a `ref.current` is set before calling `focus()` on it. I think they are always guaranteed to be set. They are accessed inside DOM handlers, which means the respective elements are always present.

Right, I tend to just write that type of safeguard mechanically without much thought. I think it's probably safe here, yes.

> * to focus the link inside the item, I created a `focusLink` instance method. There's a chain of two refs in action.

Yes, that's probably a better model than the `forwardRef` I was using.

diegohaz commented:
I'm pretty sure we loose tree shaking when doing `Dropdown.Label` instead of `DropdownLabel`. This may not be a huge problem for this component as this is really small, but I think it's worth keeping that in mind for other components.

An alternative solution that removes the deep imports and keeps tree shaking working is just re-exporting them in the index file and importing them separately from the same path:

```js
import { DropdownLabel, DropdownSeparator } from 'components/select-dropdown';
```

Regarding `ref.current`, it'll be undefined if the dom element is removed for some reason (e.g. `false && <div />`).

jsnajdr commented:
> I'm pretty sure we loose tree shaking when doing Dropdown.Label instead of DropdownLabel.

We already don't have any tree shaking for this component, because `Label` and `Separator` are used when passing an `options` array:
```jsx
<SelectDropdown options = { [
  { label: 'Unclickable', isLabel: true }, // this will be a SelectDropdown.Label
  null, // this will be a SelectDropdown.Separator
  { label: 'Real Option', value: 'opt' }, // this will be SelectDropdown.Item
] } />
```

Adding the subcomponents as static fields doesn't add any new dependencies: there already are there.

> Regarding ref.current, it'll be undefined if the dom element is removed for some reason

In our case though, the container is always rendered (no conditional rendering is there) and the `Item` refs point to the component class instance, not its DOM elements.

jsnajdr committed:
SelectDropdown: return undefined from getInitialSelectedItem

jsnajdr committed:
SelectDropdown: update unit tests

Some of them were quite awful, spying on internal method calls or mocking the whole component instance
instead of testing on the real component and checking its state and JSDOM rendering.

jsnajdr commented:
I just rebased the branch onto latest `master` (I was a bit behind) and update all the unit tests.

diegohaz commented:
> We already don't have any tree shaking for this component, because Label and Separator are used when passing an options array. Adding the subcomponents as static fields doesn't add any new dependencies: there already are there.

Ah, yeah! I missed that! I still wonder about consistency: instead of adopting `ParentChild` and `Parent.Child` for different components depending on whether they support tree shaking, isn't it better to follow one standard (`ParentChild`) since it works regardless? I don't know if it was discussed before.

I think I started checking all `ref.current`s just because TypeScript complained about it (and often times it was right).

diegohaz commented on client/components/select-dropdown/index.jsx:
```diff
@@ -48,49 +52,27 @@ class SelectDropdown extends Component {
 		style: {},
 	};
 
 	static instances = 0;
```
Not something that should be fixed in this PR or any time soon, but I think it's worth noting. If we're ever going to render this component on the server, this can potentially break since this value will be shared between all clients accessing the same server instance.

jsnajdr replied:
That's a good point: the generated IDs will be different on server and client. Thanks for linking that React issue! As most other people who commented there, all we need the ID for is to link elements together with various ARIA attributes.

I think managing IDs with React Context would solve the issue even in Async mode. I don't immediately see how Async mode could break it. Initial rendering on server and client is always sync and only the later updates can be async, preempt and cancel each other etc.

jsnajdr commented:
> I still wonder about consistency: instead of adopting `ParentChild` and `Parent.Child` for different components depending on whether they support tree shaking, isn't it better to follow one standard (`ParentChild`) since it works regardless? I don't know if it was discussed before.

I'm proposing the `Parent.Child` convention only for components that are part of the same module and are always used together, like `ul` and `li` or `select` and `option`.

The goal is to improve the DX when importing the component. Some context:

This convention is simply bad:
```jsx
import Dropdown from 'components/dropdown';
import DropdownItem from 'components/dropdown/item';
```
because it can't be easily migrated to an import from package:
```jsx
import Dropdown from '@automattic/calypso-ui/dropdown';
import DropdownItem from '@automattic/calypso-ui/dropdown/item';
```
When importing from package subdirectories like this, it's impossible to publish multiple versions of the bundle (CJS, ESM, ES:next, ...) because the `main` and `module` fields in `package.json` will be ignored.

We always need to import from the root package (where the `index.js` contains reexports and has `sideEffects: false`) and then we have two options:
```jsx
import { Dropdown, DropdownItem } from '@automattic/calypso-ui';

<Dropdown>
  <DropdownItem />
</Dropdown>
```
or
```jsx
import { Dropdown } from '@automattic/calypso-ui';

<Dropdown>
  <Dropdown.Item />
</Dropdown>
```

I like the second option much better: I import only one thing from the package. Because conceptually it's one thing with several different parts rather than two different things.

Does that make sense?

diegohaz commented:
> When importing from package subdirectories like this, it's impossible to publish multiple versions of the bundle (CJS, ESM, ES:next, ...) because the main and module fields in package.json will be ignored.

It's possible! But it requires [some additional build script](https://github.com/reakit/reakit/blob/167e6bb021dfb97cce6d10f81931fc9a30501a62/scripts/build/build.js#L25) to create [proxy directories](https://developers.livechatinc.com/blog/how-to-create-javascript-libraries-in-2018-part-2/). Those directories include only a `package.json` pointing to the canonical sources, like this:

![image](https://user-images.githubusercontent.com/3068563/62902895-62cd8d80-bd37-11e9-8ee2-8b213f0bf6ba.png)

With this approach, users are able to import either from the root or from the public subdirectories without losing anything, as explained here: https://reakit.io/docs/bundle-size/

Ideally, build tools like Babel and Webpack should be smart enough and people shouldn't be worrying about things like that (maybe [Prepack](https://prepack.io) will save us in the future?). But, until this becomes a reality, importing from subdirectories is usually useful for libraries and apps without tree shaking support. Plus, from what I've seen, people tend to adopt a library more easily if they don't feel like they're including the whole thing when they only need a piece.

Regarding importing separately or not, [I used to prefer what you're suggesting](https://github.com/reakit/reakit/blob/fd10c8bd0c55f04a36c3b8f7d949c181a8b30d96/packages/reakit/src/Hidden/index.ts#L13-L18) (`Dropdown.Item`). I agree that it has better DX. But recently I've migrated to the separate component/file since you can create bigger modules (e.g. [Form](https://github.com/reakit/reakit/tree/167e6bb021dfb97cce6d10f81931fc9a30501a62/packages/reakit/src/Form)) with child components that will not necessarily be used all the time without worrying about unnecessarily increasing bundle size.

It's up to the user if they want the convenience of importing everything as an object:
```jsx
import * as d from "reakit/Dialog";

// not pretty, but tradeoffs 🤷‍♂️
<d.DialogBackdrop />
<d.DialogDisclosure />
<d.Dialog />
```
Or if they want it separate:
```jsx
import { Dialog, DialogBackdrop } from "reakit";
import { Dialog, DialogBackdrop } from "reakit/Dialog";
import { Dialog } from "reakit/Dialog/Dialog";
import { DialogBackdrop } from "reakit/Dialog/DialogBackdrop";
```

jsnajdr commented:
> It's possible! But it requires some additional build script to create proxy directories.

Thanks, I didn't realize that! Every component can have its own `package.json` with the right pointers.

If I only converted:
```js
import SelectDropdown from 'components/select-dropdown';
import SelectDropdownItem from 'components/select-dropdown/item';
```
to
```js
import { SelectDropdown, SelectDropdownItem } from 'components/select-dropdown';
```
in this PR and avoided the controversial `SelectDropdown.Item` convention, would that be better?

The main objective of this PR is to migrate CSS, not to rearrange the imports. So I'm willing to give up that part completely if there's not 100% consensus that it's a good idea 🙂

jsnajdr commented:
> I think I started checking all `ref.currents` just because TypeScript complained about it (and often times it was right).

Consider a simple component like this:
```jsx
class ReffyComponent extends Component {
  inputRef = createRef();

  componentDidMount() {
    this.inputRef.current.focus();
  }

  render() {
    return (
      <div>
        <input type="text" ref={ this.inputRef } />
      </div>
    );
  }
}
```

At the time when `componentDidMount` is called, `this.inputRef.current` is **guaranteed** to be a valid DOM element object. At other times during the render cycle it's not always so. The ref is initially `undefined` and is set to `null` and then back to DOM object during a rerender, but at the time when `componentDidMount` or `componentDidUpdate` is called, it's always a DOM object.

We know that because we know how React works. And I understand that TypeScript doesn't know that. But in that case TypeScript becomes a liability and doesn't add value. Is there any way how to tell TypeScript that guarding `this.inputRef.current` is not necessary?

diegohaz commented:
> Is there any way how to tell TypeScript that guarding this.inputRef.current is not necessary?

We can do `this.inputRef.current!.focus()` or just `// @ts-ignore` it. I guess the problem that TS is concerned about is that the component may be something like this:
```jsx
<div>
  { loading ? 'Loading...' : <input type="text" ref={ this.inputRef } /> }
</div>
```

But, guarding it could make it harder to find the error if this component is updated to something like that in the future, so I vote for keeping it as is. :)

I think the PR is okay, I just wanted to kick off a discussion around creating a consistent pattern around modules/imports, but this can be discussed in other issue/PR.

Are the failing checks on the PR a blocker?

jsnajdr commented:
> I guess the problem that TS is concerned about is that the component may be something like this:

Yes, in that case, the ref is not always set. But TypeScript doesn't distinguish between the two cases and its advice is not very helpful.

Also, when showing a placeholder like in your example, the lifecycle method is likely to be similar to this:
```js
componentDidMount() {
  if ( this.props.loading ) {
    // do some magic on the placeholder
  } else {
    this.inputRef.current.focus();
  }
}
```
I.e., there are several variables that change together as the system moves from one state to another. There are invariants that never change:
```js
// assertion that's always true in `componentDidMount`
expect( ( loading && ! inputRef ) || ( ! loading && inputRef ) ).toBeTruthy();
```
Here, TypeScript nudges us towards refactoring the component to something where the relationship is more clear, e.g., two different components.

> I think the PR is okay, I just wanted to kick off a discussion around creating a consistent pattern around modules/imports, but this can be discussed in other issue/PR.

OK, does that mean I can keep the `SelectDropdown.Item` convention for now? It's an easy codemod if we want to change it later. (we'll need to make the final decision when moving this to `@automattic/calypso-ui`)

> Are the failing checks on the PR a blocker?

Yes, the e2e tests are consistently failing, so I likely broke something. I'll need to look into that.

diegohaz commented:
> OK, does that mean I can keep the SelectDropdown.Item convention for now? It's an easy codemod if we want to change it later. (we'll need to make the final decision when moving this to @automattic/calypso-ui)

Yes. No need to change that.

jsnajdr committed:
SelectDropdown: disable some ESLint warnings in docs example

jsnajdr committed:
SelectDropdown: hide the dropdown options when the dropdown is closed

Otherwise, the options element is clickable although it has `visibility: hidden`
and captures clicks on controls that are underneath it.

jsnajdr commented:
Found the bug that was breaking the e2e tests (namely "Revert to Draft" in Classic Editor) and fixed it in c9b3c11.

jsnajdr commented:
The remaining e2e failures are not related, let's merge 🤞
